modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/State.scala

```diff
@@ -12,50 +12,92 @@
  */
 package com.snowplowanalytics.snowplow.rdbloader
 
-import cats.effect.Concurrent
-import cats.effect.concurrent.{ Ref => CERef }
+import java.time.Instant
 
+import cats.Monad
 import cats.implicits._
+import cats.effect.{ Concurrent, Clock }
 
 import fs2.concurrent.SignallingRef
 
+import com.snowplowanalytics.snowplow.rdbloader.common.S3
+import com.snowplowanalytics.snowplow.rdbloader.loading.Load
+import com.snowplowanalytics.snowplow.rdbloader.loading.Load.State.Idle
+import com.snowplowanalytics.snowplow.rdbloader.loading.Load.State.Loading
+
 /**
- * Primary (mutable) state of the loader
+ * Primary state of the loader
  * Every Loader's action has two input parameters: message and current state
  * The state is used to exchange data between data discovery stream and load actions
  *
+ * @param loading state of the folder loading, which can be either idling ("not busy" or
+ *        "no folder") or loading at some stage
+ * @param updated when the state was updated the last time
+ *        Used to find out about degraded infra - if state is not updated for long enough
+ *        it likely means that the database is unresponsive
  * @param attempts amount of attempts the Loader took to load **current** folder
  *                 zero'ed after every message ack'ed
- * @param busy whether Loader is ready to accept new message at the moment
- *             if Loader is busy - it must be retrying until fubr, then it must
- *             set `busy` back to `false`
  * @param loaded amount of folders the loader managed to load
  * @param messages total amount of message received
  */
-case class State[F[_]](attempts: Int,
-                       busy: SignallingRef[F,  Boolean],
-                       loaded: Int,
-                       messages: Int) {
-  def incrementAttempts: State[F] =
+case class State(loading: Load.State,
+                 updated: Instant,
+                 attempts: Int,
+                 loaded: Int,
+                 messages: Int) {
+  def incrementAttempts: State =
     this.copy(attempts = attempts + 1)
-  def incrementMessages: State[F] =
+  def incrementMessages: State =
     this.copy(messages = messages + 1)
-  def incrementLoaded: State[F] =
+  def incrementLoaded: State =
     this.copy(loaded = loaded + 1)
 
+  /** Start loading a folder */
+  def start(folder: S3.Folder): State =
+    this.copy(loading = Load.State.start(folder), attempts = 0)
+  def idle: State =
+    this.copy(loading = Load.State.Idle)
+  def setStage(stage: Load.Stage): State =
+    loading match {
+      case Loading(folder, _) => this.copy(loading = Loading(folder, stage))
+      case Idle => throw new IllegalStateException(s"Cannot set $stage stage while loading is Idle")
```

Comment:
I'm flagging this, because it looks dangerous, even though I can see we never call it with Idle.  My fear is that code like this easily becomes dangerous in a future code refactor.  Is there anything we can do with types to make the illegal state impossible?

If there's something easy, then let's change it.  Otherwise, I'm happy to leave it like this for now.

Reply:
I know. My very favourite option was to add `def next: Stage` method to the `Stage` to express the finite state machine nature. Every stage has predefined next stage, but it doesn't play well with `Loading(folder)`. I think there are ways to make it work, but have their own trade-offs.